### PR TITLE
Doc objects

### PR DESCRIPTION
**Pull request by mgiammar**
Convert examples from dictionaries to objects.

### CODE REVIEW

File: src/mrsimulator/methods/base.py [hunk 1]
```diff
@@ -145,57 +145,25 @@ def update(cls, **kwargs):
...
         allowed for NamedMethods."""
-        if not isinstance(py_dict["spectral_dimensions"][0], dict):
-            return cls.check_when_arg_is_object(py_dict)
-
         default_method = cls.update(**py_dict)
         default_spectral_dimensions = default_method["spectral_dimensions"]
         for i, item in enumerate(py_dict["spectral_dimensions"]):
-            if item["events"] == [{}]:
+            if isinstance(item, SpectralDimension):
+                item = item.dict()
```
Comment (deepanshs) [inline src/mrsimulator/methods/base.py:152]: Instead of using item.dict(), use item.json(units=False). I think `dict` will return dictionary object with mrsim objects. The `json` function with return nested dictionary with no mrsim objects.

File: src/mrsimulator/methods/base.py [hunk 2]
```diff
@@ -145,57 +145,25 @@ def update(cls, **kwargs):
...
-                if k in py_keys:
-                    a1, a2, a3 = [getattr(_, k) for _ in [ev_obj, default_obj, ev_py]]
-                    a = a1 != a2 and a1 != a3 and a2 is not None
-                    setattr(ev_obj, k, a3)
-                elif k in required and k in obj_dict:
-                    a = getattr(ev_obj, k) != obj_dict[k]
-                    setattr(ev_obj, k, obj_dict[k])
-                if a:
-                    raise ImmutableEventError(cls.__name__)
-
```
Comment (deepanshs) [inline src/mrsimulator/methods/base.py:198]: @mgiammar Why did you remove these checks?

File: src/mrsimulator/methods/base.py (hunk 1 above)
Comment (mgiammar) [reply src/mrsimulator/methods/base.py:152]: `dict` adds all attributes regardless if they are defined whereas `json` only adds attributes which have been defines. Calling `json` on a SpectralDimension without any Events does not include the `events` keyword in the dictionary.

The code can be changed to call `json(units=False)` and check if `events` is present.

File: src/mrsimulator/methods/base.py (hunk 2 above)
Comment (mgiammar) [reply src/mrsimulator/methods/base.py:198]: This code was only called when a SpectralDimension was given as an object. With changing any SpectralDimension into a dictionary this code would never run (I think).

Comment (deepanshs) [reply src/mrsimulator/methods/base.py:198]: The code is called before you convert it to dict.

File: docs/introduction.rst [hunk 3]
```diff
@@ -119,42 +129,49 @@ Using the previous 2H **Site** object example, we construct a simple single-site
...
-              "beta": "1.231 rad",
-              "gamma": "3.1415 rad"
-          }
-        }
-      ],
-      "abundance": "0.148%"
-  }
+.. plot::
+    :format: doctest
+    :context: close-figs
```
Comment (deepanshs) [inline docs/introduction.rst:134]: There is no need for a `plot` directive. There are no Matplotlib figures in these examples. Just use a `doctest` directive.

File: docs/introduction.rst [hunk 4]
```diff
@@ -119,42 +129,49 @@ Using the previous 2H **Site** object example, we construct a simple single-site
...
+    ... )
 
 At the root level of the **SpinSystem** object, we find four keywords, **name**,
 **description**, **sites**, and **abundance**. The value of the *name* key is the
-optional name of the spin system, here given a value of *2H uncoupled spin system*. The
+optional name of the spin system. The
 value of the description key is an optional string describing the spin system. The
 value of the *sites* key is a list of **Site** objects. Here, this list comprises of
-single **Site** object (lines 5-22) from :numref:`listing_site`. The value of the
+a single **Site** object (lines 10-30) from :numref:`listing_site`. The value of the
```
Comment (deepanshs) [inline docs/introduction.rst:174]: There are no line number in the build. Just remove it.

File: src/mrsimulator/methods/base.py [hunk 5]
```diff
@@ -157,12 +157,16 @@ def check_method_compatibility(cls, py_dict):
...
-            if item["events"] == [{}]:
+
+            # If no methods in SpectralDimension, set to default events
+            if item["events"] == [{}] or item["events"] == []:
                 item["events"] = default_spectral_dimensions[i]["events"]
 
             elif item["events"] != default_spectral_dimensions[i]["events"]:
                 raise ImmutableEventError(cls.__name__)
 
+            py_dict["spectral_dimensions"][i] = item
```
Comment (deepanshs) [inline src/mrsimulator/methods/base.py:168]: What is the purpose of this statement? `item` is the ith spectral dimension reference from `py_dict`, so this statement is `item = item`!

File: fitting_source/1D_fitting/plot_2_13C_glycine_960Hz.py [hunk 6]
```diff
@@ -1,8 +1,8 @@
 #!/usr/bin/env python
 # -*- coding: utf-8 -*-
 """
-¹³C MAS NMR of Glycine (CSA) [960 Hz]
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+¹³C MAS NMR of Glycine (CSA) [960 Hz] (TO REMOVE)
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (deepanshs) [inline fitting_source/1D_fitting/plot_2_13C_glycine_960Hz.py:5]: Why?

File: examples_source/1D_simulation(crystalline)/plot_3_satellite_transition_sim.py [hunk 7]
```diff
@@ -116,19 +118,19 @@
...
-            "events": [
-                {
-                    "transition_query": [
+        SpectralDimension(
+            count=1024,
+            spectral_width=1e4,  # in Hz
+            reference_offset=1e4,  # in Hz
+            events=[
+                SpectralEvent(
+                    transition_query=[
```
Comment (deepanshs) [inline examples_source/1D_simulation(crystalline)/plot_3_satellite_transition_sim.py:127]: For consistency, add `"ch1"` to the query.

File: src/mrsimulator/methods/base.py (hunk 1 above)
Comment (mgiammar) [reply src/mrsimulator/methods/base.py]: These lines have been removed

File: src/mrsimulator/methods/base.py (hunk 2 above)
Comment (mgiammar) [reply src/mrsimulator/methods/base.py]: This code has been readded since it is necessary for some named method validation

File: src/mrsimulator/methods/base.py (hunk 5 above)
Comment (mgiammar) [reply src/mrsimulator/methods/base.py]: line has been removed

File: fitting_source/1D_fitting/plot_2_13C_glycine_960Hz.py (hunk 6 above)
Comment (mgiammar) [reply fitting_source/1D_fitting/plot_2_13C_glycine_960Hz.py:5]: The next example is fitting all three spinning speeds. Should an example where we fit just one spectrum still exits?

File: examples_source/1D_simulation(crystalline)/plot_3_satellite_transition_sim.py (hunk 7 above)
Comment (mgiammar) [reply examples_source/1D_simulation(crystalline)/plot_3_satellite_transition_sim.py:127]: added

File: fitting_source/1D_fitting/plot_2_13C_glycine_960Hz.py (hunk 6 above)
Comment (deepanshs) [reply fitting_source/1D_fitting/plot_2_13C_glycine_960Hz.py:5]: Yes. It should be fine. Just remove the `(To REMOVE)` text and let's have this example in the docs.

File: docs/introduction.rst (hunk 3 above)
Comment (mgiammar) [reply docs/introduction.rst:134]: Will be done in #98

File: docs/introduction.rst (hunk 4 above)
Comment (mgiammar) [reply docs/introduction.rst:174]: Will be done in #98

**Review by deepanshs — APPROVED** [review]

**Review by deepanshs — APPROVED** [review]

**Comment by deepanshs** [comment]
Took a couple of tries to have the CI run for a non master PR merge. It is ready now.
